ovos_audio/playback.py

```diff
@@ -0,0 +1,272 @@
+            self._processing_queue = False
+        # Clear cache for all attached tts objects
+        # This is basically the only safe time
+        for tts in self.attached_tts:
+            tts.cache.curate()
+        self.blink(0.2)
+
+    def _play(self):
+        try:
+            data, visemes, listen, tts_id, message = self._now_playing
+            self.activate_tts(tts_id)
+            self.on_start(message)
+
+            data = self.tts_transform.transform(data, message.context)
+
+            self.p = play_audio(data)
+            if visemes:
+                self.show_visemes(visemes)
+            if self.p:
+                self.p.communicate()
+                self.p.wait()
+            self.deactivate_tts()
+            if self.queue.empty():
+                self.on_end(listen, message)
+        except Empty:
+            pass
+        except Exception as e:
+            LOG.exception(e)
+            if self._processing_queue:
+                self.on_end()
+        self._now_playing = None
+
+    def run(self, cb=None):
+        """Thread main loop. Get audio and extra data from queue and play.
+
+        The queue messages is a tuple containing
+        snd_type: 'mp3' or 'wav' telling the loop what format the data is in
+        data: path to temporary audio data
+        videmes: list of visemes to display while playing
+        listen: if listening should be triggered at the end of the sentence.
+
+        Playback of audio is started and the visemes are sent over the bus
+        the loop then wait for the playback process to finish before starting
+        checking the next position in queue.
+
+        If the queue is empty the tts.end_audio() is called possibly triggering
+        listening.
+        """
+        self._paused = False
+        self._started.set()
+        while not self._terminated:
+            while self._paused:
+                sleep(0.2)
+            try:
+                # HACK: we do these check to account for direct usages of TTS.queue singletons
+                speech_data = self.queue.get(timeout=2)
+                if len(speech_data) == 5 and isinstance(speech_data[-1], Message):
+                    data, visemes, listen, tts_id, message = speech_data
+                else:
+                    LOG.warning("it seems you interfacing with TTS.queue directly, this is not recommended!\n"
+                                "new expected TTS.queue contents -> data, visemes, listen, tts_id, message")
+                    if len(speech_data) == 6:
+                        # old ovos backwards compat
+                        _, data, visemes, ident, listen, tts_id = speech_data
+                    elif len(speech_data) == 5:
+                        # mycroft style
+                        tts_id = None
+                        _, data, visemes, ident, listen = speech_data
+                    else:
+                        # old mycroft style  TODO can this be deprecated? its very very old
```

Comment:
Add a log message?

Reply:
leaving these deprecations for later as mentioned above

for context, this class was originally moved to OPM to allow our plugins to work in classic mycroft-core (this is no longer a goal), this was in OPM to patch the mycroft-core class. but it ends up being unrelated to the TTS plugins directly so now it comes back to the proper place

not having the compat with classic core simplifies our requirements and makes it easy to refactor this, we will be able to ditch most of these since now we are assured to get the correct data and no longer need to handle arbitrary data added to the TTS.queue

Reply:
also theres a LOG already a bit above covering all the unexpected data types :)